emission/analysis/userinput/matcher.py

```diff
@@ -38,57 +39,132 @@ def match_incoming_inputs(user_id, timerange):
+    time_query = epq.get_time_range_for_composite_object_creation(user_id)
+    ts = esta.TimeSeries.get_time_series(user_id)
+    confirmedTrips = esda.get_entries(esda.CONFIRMED_TRIP_KEY, user_id, time_query=time_query)
+    logging.info("Creating composite trips from %d confirmed trips" % len(confirmedTrips))
+    if len(confirmedTrips) == 0:
+        logging.debug("len(confirmedTrips) == 0, early return")
+        return None
+
+    last_done_ts = None
+    for ct in confirmedTrips:
+        logging.info("End place type for trip is %s" %  type(ct.data.end_place))
+        composite_trip_dict = copy.copy(ct)
+        del composite_trip_dict["_id"]
+        composite_trip_dict["metadata"]["key"] = "analysis/composite_trip"
+
+        # confirmed_trip has an id for its corresponding confirmed_place
+        # for composite_trip, we want to get the actual confirmed_place object
+        confirmed_place_id = ct["data"]["confirmed_place"]
+        confirmed_place = esda.get_entry(esda.CONFIRMED_PLACE_KEY, confirmed_place_id)
+        composite_trip_dict["data"]["confirmed_place"] = confirmed_place
+
+        # retrieve locations for the trajectory of the trip
+        time_query = estt.TimeQuery("data.ts", ct["data"]["start_ts"], ct["data"]["end_ts"])
+        locations = esda.get_entries(esda.CLEANED_LOCATION_KEY, user_id, time_query=time_query)
+        max_entries = 100; # we will downsample to 100 locations
+        if len(locations) > max_entries:
+            logging.debug('Downsampling to %d points' % max_entries)
+            sample_rate = len(locations)//max_entries + 1
+            locations = locations[::sample_rate]
+        composite_trip_dict["data"]["locations"] = locations
+
+        # later we will want to put section & modes in composite_trip as well
+
+        composite_trip_entry = ecwe.Entry(composite_trip_dict)
+        # save the entry
+        ts.insert(composite_trip_entry)
+        last_done_ts = confirmed_place["data"]["enter_ts"]
+
+    epq.mark_composite_object_creation_done(user_id, last_done_ts)
+
 def create_confirmed_objects(user_id):
     time_query = epq.get_time_range_for_confirmed_object_creation(user_id)
     try:
-        last_expected_trip_done = create_confirmed_trips(user_id, time_query)
+        # we will query the same time range for trips and places,
+        # but querying 'enter_ts' for places and 'end_ts' for trips
+        time_query.timeType = "data.enter_ts"
+        processed_places = create_confirmed_places(user_id, time_query)
+        time_query.timeType = "data.end_ts"
+        last_expected_trip_done = create_confirmed_trips(user_id, time_query, processed_places)
         if last_expected_trip_done is None:
             logging.debug("after run, last_expected_trip_done == None, must be early return")
             epq.mark_confirmed_object_creation_done(user_id, None)
         else:
             epq.mark_confirmed_object_creation_done(user_id, last_expected_trip_done.data.end_ts)
```

Comment:
So we are marking the stage as "done" with the timestamp of the last trip rather than the last place? Will we be showing places in the UI? Shouldn't the last processed be the last place timestamp then?

Reply:
Please add some comments clarifying this

Reply:
The end_ts of the last trip is always the same as the enter_ts of the last place, right?
Whichever way we access it, that seems like the right value to use.
If we used the exit_ts of the last place, then on the next pipeline run with new trips, when we query trips for start_ts, we might leave off the next trip.

Is that correct? If yes, I can add an explanatory comment.